ge/org.osate.ge/src/org/osate/ge/internal/ui/tools/ToolUtil.java

```diff
@@ -80,4 +108,234 @@ public static Context findContextExcludeOwner(final BusinessObjectContext boc,
+		final Diagnostic validationDiagnostic = Diagnostician.INSTANCE.validate(serializedObject,
+				Collections.singletonMap(Diagnostician.VALIDATE_RECURSIVELY, true));
+		if (isErrorOrWarning(validationDiagnostic.getSeverity())) {
+			// Validation Diagnostics
+			final Stream<DiagnosticWrapper> qualifiedDiagnostics = getDiagnosticDescendants(validationDiagnostic)
+					.filter(diagnostic -> diagnostic instanceof FeatureBasedDiagnostic)
+					.map(diagnostic -> diagnosticBuilder.build(diagnostic));
+			addToBuilder(diagnostics, qualifiedDiagnostics);
+		}
+
+		// Errors
+		final Stream<DiagnosticWrapper> resourceErrors = getResourceDiagnostics(resource.getErrors(),
+				diagnosticBuilder,
+				Diagnostic.ERROR);
+		addToBuilder(diagnostics, resourceErrors);
+
+		// Warnings
+		final Stream<DiagnosticWrapper> resourceWarnings = getResourceDiagnostics(resource.getWarnings(),
+				diagnosticBuilder,
+				Diagnostic.WARNING);
+		addToBuilder(diagnostics, resourceWarnings);
+
+		return diagnostics.build().collect(Collectors.toSet());
+	}
+
+	private static void addToBuilder(final Builder<DiagnosticWrapper> builder,
+			final Stream<DiagnosticWrapper> diagnosticsToAdd) {
+		diagnosticsToAdd.forEach(diagnostic -> builder.add(diagnostic));
+	}
+
+	private static class DiagnosticBuilder {
+		private final String msgPrefix;
+
+		public DiagnosticBuilder(final String msgPrefix) {
+			this.msgPrefix = msgPrefix;
+		}
+
+		public DiagnosticWrapper build(final int severity, final String msg) {
+			return new DiagnosticWrapper(severity, getDetailedDiagnosticMessage(this.msgPrefix, msg));
+		}
+
+		public DiagnosticWrapper build(final Diagnostic diagnostic) {
+			return new DiagnosticWrapper(diagnostic.getSeverity(),
+					getDetailedDiagnosticMessage(this.msgPrefix, diagnostic.getMessage()));
+		}
+
+		public DiagnosticWrapper build(final int severity,
+				final org.eclipse.emf.ecore.resource.Resource.Diagnostic diagnostic) {
+			return new DiagnosticWrapper(severity,
+					getDetailedDiagnosticMessage(this.msgPrefix, diagnostic.getMessage()));
+		}
+
+		private static String getDetailedDiagnosticMessage(final String prefix, final String message) {
+			return new StringBuilder(prefix).append(message).toString();
+		}
+	}
+
+	private static class DiagnosticWrapper extends BasicDiagnostic {
+		public DiagnosticWrapper(final int severity, final String message) {
+			this.severity = severity;
+			this.message = message;
+		}
+
+		@Override
+		public boolean equals(final Object o) {
+			if (o == null) {
+				return false;
+			}
+			if (o == this) {
+				return true;
+			}
+			if (getClass() != o.getClass()) {
+				return false;
+			}
+
+			final DiagnosticWrapper e = (DiagnosticWrapper) o;
+			return (this.message.equalsIgnoreCase(e.getMessage()));
+		}
+
+		@Override
+		public int hashCode() {
+			return this.message.hashCode();
+		}
+	}
+
+	private static boolean isErrorOrWarning(final int severity) {
```

Comment:
It would be helpful to add a javadoc indicating what `severity` is since it is just an int. It's not clear without looking into the code.

Reply:
I changed this to make sense and added javadoc